L73-java-binderchannel/wireformat.md

```diff
@@ -0,0 +1,173 @@
+# The wire format of BinderTransport.
+
+All communication occurs via one-way binder
+[transactions](https://developer.android.com/reference/android/os/Binder#transact\(int,%20android.os.Parcel,%20android.os.Parcel,%20int\))
+which are made up of an integer “code”, and a [Parcel] containing arbitrary
+data.
+
+Three types of binder are involved in our implementation.
+
+*   The _endpoint binder_, owned by the BinderServer or OnDeviceServerEndpoint
+    instance, and returned from the host Android service's onBind method.
+*   The _client binder_, owned by the ClientTransport implementation within the
+    Channel.
+*   The _server binder_, owned by the ServerTransport implementation.
+
+A single endpoint binder exists for each BinderServer or OnDeviceServerEndpoint.
+A client binder / server binder pair is created for each transport pair.
+
+## Transaction Types
+
+There are two types of transactions, control and stream.
+
+### Transport-level Control Transactions
+
+For general transport lifecycle, using one of 1000 reserved transaction codes.
+
+*   Setup Transport
+    *   Transaction code: BinderTransport.SETUP_TRANSPORT
+    *   From ClientTransport to host binder, and from ServerTransport to client
+        binder.
+*   Shutdown Transport
+    *   Transaction code: BinderTransport.SHUTDOWN_TRANSPORT
+    *   From ClientTransport to server binder, or ServerTransport to client
+        binder.
+*   Acknowledge bytes.
+    *   Transaction code: BinderTransport.ACKNOWLEDGE_BYTES
+    *   From either transport to its peer's binder. Reports reception of
+        transaction data for flow control.
+*   Ping
+    *   Transaction code: BinderTransport.PING
+    *   From ClientTransport to server binder.
+*   Ping Response:
+    *   Transaction code: BinderTransport.PING_RESPONSE
+    *   From SeverTransport to client binder, in response to a ping.
+
+#### Transaction Data
+
+```
+version = int; (* Currently always 1 *)
+num bytes = int;
+ping id = int;
+
+setup transport transaction = version, binder;
+shutdown transport transaction = nothing;
+acknowledge bytes transaction = num bytes;
+ping transaction = ping id;
+ping response transaction = ping id;
+```
+
+### Stream Transactions
+
+Aside from the reserved 1000 transaction codes, all others are used for streams.
+When the ClientTransport creates a new Stream, it assigns an ID to use for all
+transactions relating to that stream (in both directions).
+
+IDs are allocated sequentially from FIRST_CALL_ID to LAST_CALL_ID, wrapping
+around at the end. Since there are roughly 2^24 available IDs, there’s no chance
+of a collision except in pathological cases. If we _do_ detect a collision
+however, we immediately shut down the transport with an INTERNAL error code.
```

Comment:
Could we gracefully shut down the transport instead? A long-lived stream could trigger this over a period of time. Unlikely given the shorter life of Android processes, but not out-of-the-question.

Reply:
Good point, this is easy to handle gracefully as well. Updated to shutdown gracefully and fail the new call (the one which led to the collision) with UNAVAILABLE.